flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbTypeInformedElementConverter.java

```diff
@@ -0,0 +1,245 @@
+            TypeInformation<?> componentTypeInfo =
+                    objectTypeInformation instanceof BasicArrayTypeInfo
+                            ? ((BasicArrayTypeInfo<?, ?>) objectTypeInformation).getComponentInfo()
+                            : objectTypeInformation instanceof PrimitiveArrayTypeInfo
+                                    ? ((PrimitiveArrayTypeInfo<?>) objectTypeInformation)
+                                            .getComponentType()
+                                    : ((ObjectArrayTypeInfo<?, ?>) objectTypeInformation)
+                                            .getComponentInfo();
+            if (attribute instanceof Collection) {
+                return convertArrayTypeObject(
+                        ((Collection<?>) attribute).toArray(), componentTypeInfo, false);
+            }
+
+            return convertArrayTypeObject(
+                    attribute,
+                    componentTypeInfo,
+                    objectTypeInformation instanceof PrimitiveArrayTypeInfo);
+
+        } else if (objectTypeInformation instanceof TupleTypeInfo) {
+            TupleTypeInfo<?> tupleTypeInfo = (TupleTypeInfo<?>) objectTypeInformation;
+            Tuple t = (Tuple) attribute;
+            List<AttributeValue> attributeValues = new ArrayList<>();
+            for (int i = 0; i < tupleTypeInfo.getTotalFields(); i++) {
+                attributeValues.add(convertValue(t.getField(i), tupleTypeInfo.getTypeAt(i)));
+            }
+            return AttributeValue.fromL(attributeValues);
+
+        } else if (objectTypeInformation instanceof CompositeType) {
+            return AttributeValue.fromM(
+                    convertElementUsingTypeInfo(
+                            attribute, (CompositeType<attrT>) objectTypeInformation));
+
+        } else {
+            throw new IllegalArgumentException(
+                    String.format("Unsupported TypeInfo %s", objectTypeInformation));
+        }
+    }
+
+    private boolean isArrayTypeInfo(TypeInformation<?> typeInformation) {
+        return typeInformation instanceof BasicArrayTypeInfo
+                || typeInformation instanceof PrimitiveArrayTypeInfo
+                || typeInformation instanceof ObjectArrayTypeInfo;
+    }
+
+    private <attrT> AttributeValue convertArrayTypeObject(
+            Object attribute, TypeInformation<attrT> componentTypeInfo, boolean isPrimitive) {
+
+        if (componentTypeInfo.getTypeClass().isAssignableFrom(String.class)) {
+            return AttributeValue.fromSs(convertObjectArrayToStringList((Object[]) attribute));
+        } else if (componentTypeInfo instanceof NumericTypeInfo && !isPrimitive) {
+            return AttributeValue.fromNs(convertObjectArrayToStringList((Object[]) attribute));
+        } else if (componentTypeInfo instanceof NumericTypeInfo) {
+            if (attribute instanceof int[]) {
+                return AttributeValue.fromNs(
+                        (Arrays.stream((int[]) attribute)
```

Comment:
As a big fan of the stream API, it pains me to say that we [should not be using it for performance critical code](https://flink.apache.org/how-to-contribute/code-style-and-quality-java/#java-streams). Since this would be called for each record I would consider this performance critical/data-intensive. Please convert to standard loops, can extract to new methods for readability.